libs/ledger-live-common/src/hw/installLanguage.ts

```diff
@@ -0,0 +1,189 @@
+import { Observable, from, of, throwError, EMPTY } from "rxjs";
+import { catchError, concatMap, delay, mergeMap } from "rxjs/operators";
+import { DeviceOnDashboardExpected, TransportError, TransportStatusError } from "@ledgerhq/errors";
+
+import ManagerAPI from "../api/Manager";
+import { withDevice } from "./deviceAccess";
+import getDeviceInfo from "./getDeviceInfo";
+import { Language, languageIds, LanguagePackage } from "../types/languages";
+import network from "../network";
+import { LanguageInstallRefusedOnDevice } from "../errors";
+import { AppAndVersion } from "./connectApp";
+import appSupportsQuitApp from "../appSupportsQuitApp";
+import quitApp from "./quitApp";
+import getAppAndVersion from "./getAppAndVersion";
+import { isDashboardName } from "./isDashboardName";
+import Transport from "@ledgerhq/hw-transport";
+
+export type InstallLanguageEvent =
+  | {
+      type: "appDetected";
+    }
+  | {
+      type: "unresponsiveDevice";
+    }
+  | {
+      type: "progress";
+      progress: number;
+    }
+  | {
+      type: "devicePermissionRequested";
+    }
+  | {
+      type: "languageInstalled";
+    };
+
+const attemptToQuitApp = (transport, appAndVersion?: AppAndVersion): Observable<InstallLanguageEvent> =>
+  appAndVersion && appSupportsQuitApp(appAndVersion)
+    ? from(quitApp(transport)).pipe(
+        concatMap(() =>
+          of(<InstallLanguageEvent>{
+            type: "unresponsiveDevice",
+          })
+        ),
+        catchError((e) => throwError(e))
+      )
+    : of({
+        type: "appDetected",
+      });
+
+export type InstallLanguageRequest = {
+  deviceId: string;
+  language: Language;
+};
+
+export default function installLanguage({
+  deviceId,
+  language,
+}: InstallLanguageRequest): Observable<InstallLanguageEvent> {
+  const sub = withDevice(deviceId)(
+    (transport) =>
+      new Observable<InstallLanguageEvent>((subscriber) => {
+        const timeoutSub = of<InstallLanguageEvent>({
+          type: "unresponsiveDevice",
+        })
+          .pipe(delay(1000))
+          .subscribe((e) => subscriber.next(e));
+
+        const sub = from(getDeviceInfo(transport))
+          .pipe(
+            mergeMap(async (deviceInfo) => {
+              timeoutSub.unsubscribe();
+
+              if (language === "english") {
+                await uninstallAllLanguages(transport);
+                subscriber.next({
+                  type: "languageInstalled",
+                });
+                subscriber.complete();
+                return;
+              }
+
+              const languages = await ManagerAPI.getLanguagePackagesForDevice(deviceInfo);
+
```

Comment:
I don't understand your double vs single new line spacing strategy, I'm intrigued

Reply:
It's usually very much on the feeling. But this one does look weird, I think I'd add another line after 74 and after 78

Reply:
sure!